### PR TITLE
Formatted String Literals

### LINKED ISSUE #4444: String interpolation clash causing named capture group fail in regexp_replace()
While verifying the changes for #4296, I spotted what I think may be a bug. Repro is with Zed commit 5b40e33.

I saw in the reviews of #4435 that we're using functionality described at https://pkg.go.dev/regexp#Regexp.Expand. Therefore I tried to use the example shown there in my verification. The following baseline shows it working much as it's shown in the Go docs.

```
$ zq -version
Version: v1.6.0-9-g5b40e33b

$ cat lines.txt 
# comment line
option1: value1
option2: value2

# another comment line
option3: value3

$ zq -i line lines.txt 
"# comment line"
"option1: value1"
"option2: value2"
""
"# another comment line"
"option3: value3"

$ zq -i line lines.txt | zq -z 'regexp_replace(this,/(?m)(?P<key>\w+):\s+(?P<value>\w+)$/,"$key=$value")' -
"# comment line"
"option1=value1"
"option2=value2"
null(string)
"# another comment line"
"option3=value3"
```

However, the docs also indicate the syntax `${key}` should work, and I find it doesn't.

```
$ zq -i line lines.txt | zq -z 'regexp_replace(this,/(?m)(?P<key>\w+):\s+(?P<value>\w+)$/,"${key}=${value}")' -
error("regexp_replace: string values required for all args")
error("regexp_replace: string values required for all args")
error("regexp_replace: string values required for all args")
error("regexp_replace: string values required for all args")
error("regexp_replace: string values required for all args")
error("regexp_replace: string values required for all args")
```

I think it's a clash with Zed's [string interpolation](https://zed.brimdata.io/docs/language/overview#7111-string-interpolation). 🙁

I sense it's needed in cases where the variable reference could be ambiguous. For instance, if I wanted the key/value output to be bookended by `FOO` and `BAR` and tried to write it without the `{}`, I don't get the `BAR`.

```
$ zq -i line lines.txt | zq -z 'regexp_replace(this,/(?m)(?P<key>\w+):\s+(?P<value>\w+)$/,"FOO$key=$valueBAR")' -
"# comment line"
"FOOoption1="
"FOOoption2="
null(string)
"# another comment line"
"FOOoption3="
```

As we're fixing this, it seems it would be good to add a test of named capture groups and an example in the docs that shows them.

### PR DESCRIPTION
**Pull request by mattnibs**
The commit removes the template literals and replaces them with formatted string literals. F-strings have a similar functionality but the expression substitution only happens on strings prefixed with the 'f' character.

Fixes #4444

### CODE REVIEW

File: docs/language/expressions.md [hunk 1]
```diff
@@ -270,38 +270,39 @@ must conform to UTF-8 encoding and follow the JavaScript escaping
 conventions and unicode escape syntax.  Also, if the sequence `${` appears
 in a string the `$` character must be escaped, i.e., `\$`.
 
-### String Interpolation
+### Formatted String Literals
 
-Strings may include interpolation expressions, which has the form
+A formatted string literal (or f-string) is a string literal prefixed with 'f'.
```
Comment (philrz) [inline docs/language/expressions.md:275]:
```suggestion
A formatted string literal (or f-string) is a string literal prefixed with `f`.
```

**Review by philrz — COMMENTED** [review]
I looked over all the docs changes and made one small suggestion but overall looks good! 👍

**Review by nwt — COMMENTED** [review]
`parser.peg` at eb5839683024b3d9d55c404107e5a8aceba52ad9 is broken for me. Is it broken for you as well?

```console
$ make -C compiler/parser
/Users/noah/brimdata/zed/compiler/parser/deps/bin/pigeon -support-left-recursion -o parser.go parser.peg
format error:  filename:15114:53: missing ',' before newline in argument list (and 10 more errors)
make: *** [parser.go] Error 6
```

**Comment by mattnibs** [comment]
> `parser.peg` at [eb58396](https://github.com/brimdata/zed/commit/eb5839683024b3d9d55c404107e5a8aceba52ad9) is broken for me. Is it broken for you as well?
> 
> ```
> $ make -C compiler/parser
> /Users/noah/brimdata/zed/compiler/parser/deps/bin/pigeon -support-left-recursion -o parser.go parser.peg
> format error:  filename:15114:53: missing ',' before newline in argument list (and 10 more errors)
> make: *** [parser.go] Error 6
> ```

You are right! Will fix...

File: compiler/parser/parser.peg [hunk 2]
```diff
@@ -1289,49 +1290,53 @@ ComplexType
...
-    }
+StringLiteral
+  = '"' v:DoubleQuotedChar* '"' { return newPrimitive(c, "string", joinChars(v), nil }
+  / "'" v:SingleQuotedChar* "'" { return newPrimitive(c, "string", joinChars(V), nil }
 
-TemplateLiteralParts
-  = '"' v:TemplateDoubleQuotedPart* '"' { return v, nil }
-  / "'" v:TemplateSingleQuotedPart* "'" { return v, nil }
+FString
+  = "f\"" v:FStringDoubleQuotedElem* '"' {
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply compiler/parser/parser.peg:1298]: This gets a pigeon parser error- I think it expects a single quote to be a character.

File: compiler/ast/ast.go [hunk 3]
```diff
@@ -297,6 +297,47 @@ type OverExpr struct {
...
+}
+
+func (f *FStringText) Pos() int { return f.TextPos }
+func (f *FStringText) End() int { return f.TextPos + len(f.Text) }
+
+type FStringExpr struct {
+	Kind   string `json:"kind" unpack:""`
+	Lbrace int    `json:"lbrack"`
+	Expr   Expr   `json:"expr"`
+	Rbrace int    `json:"rbrack"`
```
Comment (nwt) [inline compiler/ast/ast.go:332]:
```suggestion
	Lbrace int    `json:"lbrace"`
	Expr   Expr   `json:"expr"`
	Rbrace int    `json:"rbrace"`
```

File: zfmt/ast.go [hunk 4]
```diff
@@ -207,6 +207,21 @@ func (c *canon) expr(e ast.Expr, parent string) {
...
+		for _, elem := range e.Elems {
+			switch elem := elem.(type) {
+			case *ast.FStringExpr:
+				c.write("{")
+				c.expr(elem.Expr, "")
+				c.write("}")
+			case *ast.FStringText:
+				c.write(elem.Text)
+			default:
+				c.write("(unknown f-strint element %T)", elem)
```
Comment (nwt) [inline zfmt/ast.go:221]:
```suggestion
				c.write("(unknown f-string element %T)", elem)
```

**Review by nwt — APPROVED** [review]
:shipit: after fixing those last few typos.

**Pull request merged**